File: judo-sdk/src/main/java/com/judopay/view/CardNumberEntryView.java
package com.judopay.view;

import android.content.Context;
import android.content.res.Resources;
import android.support.design.widget.TextInputLayout;
import android.text.InputFilter;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.judopay.R;
import com.judopay.model.CardNetwork;
import com.judopay.model.CardToken;
import com.judopay.validation.Validation;

import static com.judopay.model.CardNetwork.AMEX;

/**
 * A view that allows for card number data to be input by the user and the detected card type
 * to be displayed alongside the card number.
 * Does not perform validation itself, this is done by the {@link com.judopay.validation.CardNumberValidator}
 * class.
 */
public class CardNumberEntryView extends RelativeLayout {

    private EditText editText;
    private TextInputLayout inputLayout;
    private View scanCardButton;
    private CardTypeImageView cardTypeImageView;
    private NumberFormatTextWatcher numberFormatTextWatcher;

    public interface ScanCardButtonListener {
        void onClick();
    }

    public CardNumberEntryView(Context context) {
        super(context);
        initialize(context);
    }

    public CardNumberEntryView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initialize(context);
    }

    public CardNumberEntryView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        initialize(context);
    }

    private void initialize(Context context) {
        LayoutInflater inflater = LayoutInflater.from(context);
        inflater.inflate(R.layout.view_card_number_entry, this);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        this.cardTypeImageView = (CardTypeImageView) findViewById(R.id.card_type_view);
        TextView cardNumberHelperText = (TextView) findViewById(R.id.card_number_helper_text);
        this.editText = (EditText) findViewById(R.id.card_number_edit_text);
        this.inputLayout = (TextInputLayout) findViewById(R.id.card_number_input_layout);
        this.scanCardButton = findViewById(R.id.scan_card_button);

        editText.setOnFocusChangeListener(new CompositeOnFocusChangeListener(
                new EmptyTextHintOnFocusChangeListener(cardNumberHelperText),
                new ViewAlphaChangingTextWatcher(editText, cardTypeImageView),
                new HintFocusListener(editText, getResources().getString(R.string.card_number_format))
        ));

        numberFormatTextWatcher = new NumberFormatTextWatcher(editText, getResources().getString(R.string.card_number_format));
        editText.addTextChangedListener(numberFormatTextWatcher);
        editText.addTextChangedListener(new HidingViewTextWatcher(cardNumberHelperText));
    }

    public void setScanCardListener(final ScanCardButtonListener listener) {
        if (listener != null) {
            this.scanCardButton.setVisibility(VISIBLE);
            this.cardTypeImageView.setVisibility(GONE);

            this.scanCardButton.setOnClickListener(new SingleClickOnClickListener() {
                @Override
                public void doClick() {
                    listener.onClick();
                }
            });
        } else {
            this.scanCardButton.setVisibility(GONE);
            this.cardTypeImageView.setVisibility(VISIBLE);
            this.scanCardButton.setOnClickListener(null);
        }
    }

    public void setCardType(int type, boolean animate) {
        cardTypeImageView.setCardType(type, animate);
        if (type != 0) {
            setScanCardListener(null);
        }

        switch (type) {
            case CardNetwork.AMEX:
                setMaxLength(17);
                numberFormatTextWatcher.setFormat(getResources().getString(R.string.amex_card_number_format));
                break;

            default:
                setMaxLength(19);
                numberFormatTextWatcher.setFormat(getResources().getString(R.string.card_number_format));
        }
    }

    public void setText(String text) {
        editText.setText(text);
    }

    private void setMaxLength(int maxLength) {
        editText.setFilters(new InputFilter[]{new InputFilter.LengthFilter(maxLength)});
    }

    public int getCardType() {
        return CardNetwork.fromCardNumber(editText.getText().toString());
    }

    public String getText() {
        return editText.getText().toString().replaceAll(" ", "");
    }

    public void addTextChangedListener(SimpleTextWatcher watcher) {
        editText.addTextChangedListener(watcher);
    }

    public void setTokenCard(CardToken cardToken) {
        editText.setEnabled(false);
        boolean amex = cardToken.getType() == AMEX;

        editText.removeTextChangedListener(numberFormatTextWatcher);
        editText.setText(getContext().getString(amex ? R.string.amex_token_card_number : R.string.token_card_number, cardToken.getLastFour()));
        editText.addTextChangedListener(numberFormatTextWatcher);

        cardTypeImageView.setAlpha(1.0f);
    }

    public void setValidation(Validation validation) {
        inputLayout.setErrorEnabled(validation.isShowError());

        if (validation.isShowError()) {
            inputLayout.setError(getResources().getString(validation.getError()));
        } else {
            inputLayout.setError("");
        }
    }

    public EditText getEditText() {
        return editText;
    }

}

File: judo-sdk/src/main/java/com/judopay/view/CardNumberEntryView.kt
package com.judopay.view

import android.content.Context
import android.os.Bundle
import android.os.Parcelable
import android.text.InputFilter
import android.text.InputFilter.LengthFilter
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import androidx.constraintlayout.widget.ConstraintSet
import androidx.core.content.ContextCompat
import com.judopay.R
import com.judopay.model.CardNetwork
import com.judopay.model.CardToken
import com.judopay.validation.Validation
import kotlinx.android.synthetic.main.view_card_number_entry.view.cardNumberContainer
import kotlinx.android.synthetic.main.view_card_number_entry.view.cardNumberEditText
import kotlinx.android.synthetic.main.view_card_number_entry.view.cardNumberError
import kotlinx.android.synthetic.main.view_card_number_entry.view.cardNumberImageView

/**
 * A view that allows for card number data to be input by the user and the detected card type
 * to be displayed alongside the card number.
 * Does not perform validation itself, this is done by the {@link com.judopay.validation.CardNumberValidator}
 * class.
 */
class CardNumberEntryView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : FrameLayout(context, attrs, defStyle) {

    private lateinit var numberFormatTextWatcher: NumberFormatTextWatcher
    private var cardType = 0

    init {
        LayoutInflater.from(context).inflate(R.layout.view_card_number_entry, this)
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        numberFormatTextWatcher =
            NumberFormatTextWatcher(
                cardNumberEditText,
                resources.getString(R.string.card_number_format)
            )
        cardNumberEditText.addTextChangedListener(numberFormatTextWatcher)
    }

    override fun onSaveInstanceState(): Parcelable? = Bundle().apply {
        putParcelable(KEY_SUPER_STATE, super.onSaveInstanceState())
        putInt(KEY_CARD_TYPE, cardType)
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        if (state is Bundle) {
            val superState = state.getParcelable<Parcelable>(KEY_SUPER_STATE)
            setCardType(state.getInt(KEY_CARD_TYPE), false)
            super.onRestoreInstanceState(superState)
        } else {
            super.onRestoreInstanceState(state)
        }
    }

    fun setCardType(type: Int, animate: Boolean) {
        cardType = type
        cardNumberImageView.setImageType(type, animate)
        when (type) {
            CardNetwork.AMEX -> {
                setMaxLength(17)
                numberFormatTextWatcher.setFormat(resources.getString(R.string.amex_card_number_format))
            }
            CardNetwork.DINERS_CLUB_INTERNATIONAL -> {
                setMaxLength(16)
                numberFormatTextWatcher.setFormat(resources.getString(R.string.diners_club_international_card_number_format))
            }
            else -> {
                setMaxLength(19)
                numberFormatTextWatcher.setFormat(resources.getString(R.string.card_number_format))

            }
        }
    }

    fun setText(text: String?) {
        cardNumberEditText.setText(text)
    }

    private fun setMaxLength(maxLength: Int) {
        cardNumberEditText.filters = arrayOf<InputFilter>(LengthFilter(maxLength))
    }

    fun getCardType(): Int = CardNetwork.fromCardNumber(cardNumberEditText.text.toString())

    fun getText(): String? = cardNumberEditText.text.toString().replace(" ".toRegex(), "")

    fun addTextChangedListener(watcher: SimpleTextWatcher?) {
        cardNumberEditText.addTextChangedListener(watcher)
    }

    fun setTokenCard(cardToken: CardToken) {
        val amex = cardToken.type == CardNetwork.AMEX
        cardNumberEditText.apply {
            isEnabled = false
            removeTextChangedListener(numberFormatTextWatcher)
            setText(
                context.getString(
                    if (amex) R.string.amex_token_card_number else R.string.token_card_number,
                    cardToken.lastFour
                )
            )
            addTextChangedListener(numberFormatTextWatcher)
        }
    }

    fun setValidation(validation: Validation) {
        val set = ConstraintSet()
        set.clone(cardNumberContainer)
        if (validation.isShowError) {
            set.apply {
                clear(cardNumberEditText.id, ConstraintSet.TOP)
                setMargin(cardNumberEditText.id, ConstraintSet.BOTTOM, 6)
            }
            cardNumberEditText.apply {
                visibility = View.VISIBLE
                setTextColor(ContextCompat.getColor(context, R.color.error))
            }
            cardNumberError.setText(validation.error)
        } else {
            set.apply {
                connect(
                    cardNumberEditText.id,
                    ConstraintSet.TOP,
                    ConstraintSet.PARENT_ID,
                    ConstraintSet.TOP
                )
                setMargin(cardNumberEditText.id, ConstraintSet.BOTTOM, 0)
            }
            cardNumberEditText.setTextColor(ContextCompat.getColor(context, R.color.black))
            cardNumberError.apply {
                visibility = View.GONE
                text = ""
            }
        }
        set.applyTo(cardNumberContainer)
    }

    fun getEditText(): JudoEditText = cardNumberEditText
}